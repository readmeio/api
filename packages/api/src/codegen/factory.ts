import type CodeGenerator from './codegenerator.js';
import type Storage from '../storage.js';
import type Oas from 'oas';

import TSGenerator from './languages/typescript/index.js';

export enum SupportedLanguages {
  JS = 'js',
}
export interface InstallerOptions {
  /**
   * Will initiate a dry run install process. Used for simulating installations within a unit test.
   */
  dryRun?: boolean;

  /**
   * Used for stubbing out the logger that we use within the installation process so it can be
   * easily introspected without having to mock out `console.*`.
   */
  logger?: (msg: string) => void;
}

export function codegenFactory(
  language: SupportedLanguages,
  spec: Oas,
  specPath: string,
  identifier: string,
): CodeGenerator {
  switch (language) {
    case SupportedLanguages.JS:
      return new TSGenerator(spec, specPath, identifier);

    default:
      throw new TypeError(`Unsupported language supplied: ${language}`);
  }
}

export function uninstallerFactory(
  language: SupportedLanguages,
  storage: Storage,
  opts: InstallerOptions = {},
): Promise<void> {
  switch (language) {
    case SupportedLanguages.JS:
      return TSGenerator.uninstall(storage, opts);

    default:
      throw new TypeError(`Unsupported language supplied: ${language}`);
  }
}
