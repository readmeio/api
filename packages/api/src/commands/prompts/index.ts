import promptTerminal from '../../lib/prompt.js';
import Storage from '../../storage.js';

export async function confirmProposedIdentifier(initial: string, proposed: string) {
  return promptTerminal({
    type: 'confirm',
    name: 'value',
    message: `Looks like this API is called "${initial}", would you like to install it as "@api/${proposed}"?`,
  });
}

export async function promptForIdentifier(sdkExplanation: boolean = true) {
  let message = 'What would you like to identify this API as?';
  if (sdkExplanation) {
    message += ' This will be how you use the SDK.';
  }

  message += ' (e.g. entering `petstore` would result in `@api/petstore`)';

  return promptTerminal({
    type: 'text',
    name: 'value',
    message,
    validate: value => {
      if (!value) {
        return false;
      }

      try {
        return Storage.isIdentifierValid(value, true);
      } catch (err) {
        return err.message;
      }
    },
  });
}
