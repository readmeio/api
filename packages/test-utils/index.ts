export * from './fetch-mock.js';
export * from './load-spec.js';
