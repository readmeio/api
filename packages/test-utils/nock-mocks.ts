import type nock from 'nock';

import DatauriParser from 'datauri/parser';

export const response = { id: 123, name: 'Buster' };

export const responses = {
  all(this: nock.ReplyFnContext, uri: string, payload: nock.Body) {
    return [200, { uri, headers: this.req.headers, requestBody: payload }];
  },

  datauri(this: nock.ReplyFnContext, uri: string, payload: nock.Body) {
    const buffer = Buffer.from(payload as string, 'hex');
    const parser = new DatauriParser();

    return [
      200,
      {
        uri,
        requestBody: parser.format('.png', buffer).content,
        headers: this.req.headers,
      },
    ];
  },

  delay: (res: Record<string, unknown> | string, delay: number) => {
    return new Promise(resolve => {
      setTimeout(() => resolve(res), delay);
    });
  },

  headers: function responseHeaders(this: nock.ReplyFnContext) {
    return [200, this.req.headers];
  },

  multipart: function responseMultipart(this: nock.ReplyFnContext, uri: string, body: nock.Body): nock.ReplyFnResult {
    return [200, { uri, requestBody: body, headers: this.req.headers, boundary: body.split('\r\n')[0] }];
  },

  real: (payload: nock.Body) => {
    return () => [200, payload];
  },

  requestBody: (url: string, payload: nock.Body) => {
    return [200, { uri: url, requestBody: payload }];
  },

  searchParams: (uri: string) => {
    const res = new URL(`https://example.com${uri}`);
    return [200, `${res.pathname}${res.search}`];
  },

  url: (prop: keyof URL) => {
    return (uri: string) => {
      return [200, new URL(`https://example.com${uri}`)[prop]];
    };
  },
};
