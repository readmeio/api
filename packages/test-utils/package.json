{
  "private": true,
  "name": "@api/test-utils",
  "version": "7.0.0-alpha.0",
  "scripts": {
    "lint:types": "tsc --noEmit"
  },
  "dependencies": {
    "caseless": "^0.12.0",
    "datauri": "^4.1.0",
    "formdata-to-string": "^1.0.0",
    "vitest": "^0.34.5"
  },
  "devDependencies": {
    "@types/caseless": "^0.12.3",
    "oas": "^21.1.1",
    "typescript": "^5.2.2"
  },
  "prettier": "@readme/eslint-config/prettier"
}
