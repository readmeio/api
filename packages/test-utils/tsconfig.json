{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "lib": ["DOM", "ES2023"],
    "noEmit": true,
    "resolveJsonModule": true
  },
  "include": ["./**/*"]
}
