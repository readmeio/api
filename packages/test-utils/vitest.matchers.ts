import type { ParameterObject } from 'oas/rmoas.types';

import caseless from 'caseless';
import { expect } from 'vitest';

interface CustomMatchers<R = unknown> {
  /**
   * Assert that a given array within an OpenAPI definition has been dereferenced.
   */
  toBeDereferenced(): R;

  /**
   * Determine if a given `Headers` object has a given header matching a specific value.
   *
   * @example <caption>should match a value</caption>
   * expect(request.headers).to.have.header('connection', 'close');
   *
   * @example <caption>should match a regex</caption>
   * expect(response.headers).to.have.header('content-type', /application\/json(;\s?charset=utf-8)?/);
   *
   * @example <caption>should match one of many values</caption>
   * expect(request.headers).to.have.header('connection', ['close', 'keep-alive']);
   *
   */
  toHaveHeader(header: string, expected: RegExp | (string | number)[] | string): R;
}

declare module 'vitest' {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  interface Assertion<T = any> extends CustomMatchers<T> {}
  interface AsymmetricMatchersContaining extends CustomMatchers {}
}

expect.extend({
  toBeDereferenced(spec: ParameterObject[]) {
    const pass = !spec.filter(obj => '$ref' in obj).length;
    if (!pass) {
      return {
        message: () => 'expected the spec to be dereferenced',
        pass: false,
      };
    }

    return {
      message: () => 'expected the spec to not be dereferenced',
      pass: true,
    };
  },

  toHaveHeader(obj: Headers, header: string, expected: RegExp | (string | number)[] | string) {
    const headers = caseless(Object.fromEntries(Array.from(obj.entries())));

    // Header value should match a given regex.
    if (expected instanceof RegExp) {
      if (!expected.test(headers.get(header))) {
        return {
          message: () => `expected header to match ${expected.source}\n\nreceived: ${header}`,
          pass: false,
        };
      }

      return {
        message: () => `expected header to not match ${expected.source}\n\nreceived: ${header}`,
        pass: true,
      };
    }

    // Header value should exist in a given list.
    if (Array.isArray(expected)) {
      if (!expected.some(h => h === headers.get(header))) {
        return {
          message: () => `expected header to be one of the following: ${expected}\n\nreceived: ${header}`,
          pass: false,
        };
      }

      return {
        message: () => `expected header to not be one of the following: ${expected}\n\nreceived: ${header}`,
        pass: true,
      };
    }

    // Header value should match a given value.
    if (headers.get(header) !== expected) {
      return {
        message: () => `expected header to be ${expected}\n\nreceived: ${header}`,
        pass: false,
      };
    }

    return {
      message: () => `expected header not to be ${expected}\n\nreceived: ${header}`,
      pass: true,
    };
  },
});
