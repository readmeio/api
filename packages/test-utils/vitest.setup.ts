import nock from 'nock';
import { beforeAll, afterAll } from 'vitest';

beforeAll(() => {
  nock.disableNetConnect();
});

afterAll(() => {
  nock.cleanAll();
  nock.enableNetConnect();
});
