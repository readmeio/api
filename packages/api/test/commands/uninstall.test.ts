import type { SpyInstance } from 'vitest';

import { loadSpec } from '@api/test-utils';
import { CommanderError } from 'commander';
import fetchMock from 'fetch-mock';
import prompts from 'prompts';
import uniqueTempDir from 'unique-temp-dir';
import { describe, beforeEach, it, expect, vi, afterEach } from 'vitest';

import { SupportedLanguages } from '../../src/codegen/factory.js';
import * as codegenFactoryModule from '../../src/codegen/factory.js';
import installCmd from '../../src/commands/uninstall.js';
import Storage from '../../src/storage.js';

const baseCommand = ['api', 'uninstall'];

const cmdError = (msg: string) => new CommanderError(0, '', msg);

describe('install command', () => {
  let stdout: string[];
  let stderr: string[];
  let consoleLogSpy: SpyInstance;

  const getCommandOutput = () => {
    return [consoleLogSpy.mock.calls.join('\n\n')].filter(Boolean).join('\n\n');
  };

  beforeEach(() => {
    stdout = [];
    stderr = [];

    installCmd.exitOverride();
    installCmd.configureOutput({
      writeOut: str => stdout.push(str),
      writeErr: str => stderr.push(str),
    });

    consoleLogSpy = vi.spyOn(console, 'log').mockImplementation(() => {});
    Storage.setStorageDir(uniqueTempDir());
  });

  afterEach(() => {
    fetchMock.restore();
    Storage.reset();
    vi.restoreAllMocks();
  });

  it('should error out if identifier is not passed', () => {
    return expect(installCmd.parseAsync([...baseCommand])).rejects.toStrictEqual(
      cmdError("error: missing required argument 'identifier'"),
    );
  });

  it('should error out if invalid identifier is passed', () => {
    return expect(installCmd.parseAsync([...baseCommand, 'non-existent-identifier'])).rejects.toStrictEqual(
      new Error(
        'You do not appear to have non-existent-identifier installed. You can run `npx api list` to see what SDKs are present.',
      ),
    );
  });

  it('should print help screen', async () => {
    await expect(installCmd.parseAsync([...baseCommand, '--help'])).rejects.toStrictEqual(cmdError('(outputHelp)'));

    expect(stdout.join('\n')).toMatchSnapshot();
  });

  it('should successfully uninstall SDK', async () => {
    prompts.inject([true]);

    const petstoreSimple = await loadSpec('@readme/oas-examples/3.0/json/petstore-simple.json');
    fetchMock.get('https://dash.readme.com/api/v1/api-registry/n6kvf10vakpemvplx', petstoreSimple);

    const source = '@petstore/v1.0#n6kvf10vakpemvplx';
    const identifier = 'petstore-to-be-uninstalled';
    const storage = new Storage(source, SupportedLanguages.JS, identifier);

    await storage.load();

    expect(Storage.getLockfile().apis).toHaveLength(1);

    const uninstallSpy = vi.spyOn(codegenFactoryModule, 'uninstallerFactory').mockResolvedValue();

    await expect(installCmd.parseAsync([...baseCommand, identifier])).resolves.toBeDefined();
    expect(uninstallSpy).toHaveBeenCalledTimes(1);
    expect(getCommandOutput()).toBe('ðŸš€ All done!');
    expect(Storage.getLockfile().apis).toHaveLength(0);
  });

  it('should uninstall SDK and bypass prompt with --yes option', async () => {
    const petstoreSimple = await loadSpec('@readme/oas-examples/3.0/json/petstore-simple.json');
    fetchMock.get('https://dash.readme.com/api/v1/api-registry/n6kvf10vakpemvplx', petstoreSimple);

    const source = '@petstore/v1.0#n6kvf10vakpemvplx';
    const identifier = 'petstore-to-be-uninstalled';
    const storage = new Storage(source, SupportedLanguages.JS, identifier);

    await storage.load();

    expect(Storage.getLockfile().apis).toHaveLength(1);

    const uninstallSpy = vi.spyOn(codegenFactoryModule, 'uninstallerFactory').mockResolvedValue();

    await expect(installCmd.parseAsync([...baseCommand, identifier, '--yes'])).resolves.toBeDefined();
    expect(uninstallSpy).toHaveBeenCalledTimes(1);
    expect(getCommandOutput()).toBe('ðŸš€ All done!');
    expect(Storage.getLockfile().apis).toHaveLength(0);
  });
});
