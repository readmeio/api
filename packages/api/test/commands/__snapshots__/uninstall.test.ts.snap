// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`install command > should print help screen 1`] = `
"Usage: uninstall [options] <identifier>

uninstall an SDK from your codebase

Arguments:
  identifier  the SDK to uninstall

Options:
  -y, --yes   Automatically answer "yes" to any prompts printed
  -h, --help  display help for command


Examples:
  $ npx api uninstall petstore
"
`;
